packages/hub/test/session.test.ts

```diff
@@ -0,0 +1,117 @@
+import { SessionContext, SessionStatus } from '@sentry/types';
+import { timestampInSeconds } from '@sentry/utils';
+
+import { Session } from '../src/session';
+
+describe('Session', () => {
+  it('initializes with the proper defaults', () => {
+    const session = new Session().toJSON();
+
+    const sessionStartTime = session.timestamp;
+
+    expect(session).toEqual({
+      attrs: {},
+      duration: 0,
+      errors: 0,
+      init: true,
+      sid: expect.any(String),
+      started: expect.any(String),
+      status: SessionStatus.Ok,
+      timestamp: expect.any(String),
+    });
+
+    expect(session.sid).toHaveLength(32);
+
+    // started and timestamp should be the same on creation
+    expect(session.started).toEqual(sessionStartTime);
+    expect(session.timestamp).toEqual(sessionStartTime);
+  });
+
+  describe('update', () => {
+    const time = timestampInSeconds();
+    // [ name, in, out ]
+    const table: Array<[string, SessionContext, Record<string, any>]> = [
+      ['sets an ip address', { user: { ip_address: '0.0.0.0' } }, { attrs: { ip_address: '0.0.0.0' } }],
+      ['sets a did', { user: { id: 'specialID123' } }, { did: 'specialID123' }],
+      ['sets a timestamp', { timestamp: time }, { timestamp: new Date(time * 1000).toISOString() }],
+      ['sets a sid', { sid: '99705f22a3f1468e95ba7386e84691aa' }, { sid: '99705f22a3f1468e95ba7386e84691aa' }],
+      ['requires custom sid to be of certain length', { sid: '123' }, { sid: expect.not.stringMatching('123') }],
+      ['requires custom sid to be of certain length', { sid: '123' }, { sid: expect.not.stringMatching('123') }],
+      ['sets an init', { init: false }, { init: false }],
+      ['sets an did', { did: 'specialID123' }, { did: 'specialID123' }],
+      ['overwrites user did with custom did', { did: 'custom-did', user: { id: 'user-id' } }, { did: 'custom-did' }],
+      ['sets a started time', { started: time }, { started: new Date(time * 1000).toISOString() }],
+      ['does not set a duration for browser env', { isBrowser: true }, { duration: undefined }],
+      ['sets a duration', { duration: 12000 }, { duration: 12000 }],
+      ['does not use custom duration for browser env', { duration: 12000, isBrowser: true }, { duration: undefined }],
+      [
+        'does not set a negative duration',
+        { timestamp: 10, started: 100 },
+        // TODO(abhi): What should the behaviour here be?
+        { duration: 0, timestamp: expect.any(String), started: expect.any(String) },
```

Comment:
I think if the caller passes all three `duration`, `timestamp` and `started` we should just use those without changes.

In practice, I think `timestamp` should be computed internally as it is defined as the time when the session is sent out (so it should be set as close to delivery time as possible), and `started` is also automatically computed internally on instantiation, and, finally, `duration` is derived from the other two.

In other words, our own SDK code should not be passing any of those options in calls to `session.update`.